lib/Interpreter/IncrementalCUDADeviceCompiler.h

```diff
@@ -157,49 +116,78 @@ namespace cling {
-                   const clang::codegenoptions::DebugInfoKind debugInfo);
-
-    ///\brief Save .cu file, if cuda device code compiler failed at translation.
-    ///
-    ///\returns The error code of the rename operation
-    std::error_code saveFaultyCUfile();
+                   const clang::codegenoptions::DebugInfoKind debugInfo,
+                   const llvm::Triple hostTriple);
 
   public:
     ///\brief Constructor for IncrementalCUDADeviceCompiler
     ///
-    ///\param [in] filePath - All files will generated in the folder of the
-    ///       filePath, except the fatbin file, if it have another path. Have
-    ///       to end with a separator. Can be empty.
+    ///\param [in] fileBuffer - buffer in which the generated code of the
+    ///       transaction is stored
     ///\param [in] optLevel - The optimization level of the interpreter
-    /// instance.
+    ///       instance.
     ///       The value will be copied, because a change of it is not allowed.
     ///\param [in] invocationOptions - Contains values for the arguments of
-    ///       clang and the NVIDIA tool fatbinary.
+    ///       the interpreter instance and the NVIDIA tool fatbinary.
     ///\param [in] CI - Will be used for m_CuArgs and the include path handling.
     IncrementalCUDADeviceCompiler(
-        const std::string& filePath, const int optLevel,
-        const cling::InvocationOptions& invocationOptions,
+        std::shared_ptr<llvm::SmallVectorImpl<char>> fatbinBuffer,
+        const int optLevel, const cling::InvocationOptions& invocationOptions,
         const clang::CompilerInstance& CI);
 
+    ///\brief Returns a reference to the PTX interpreter
+    ///
+    ///\return std::unique_ptr< cling::Interpreter >&
+    ///
+    std::unique_ptr<Interpreter>& getPTXInterpreter() { return m_PTX_interp; }
+
     ///\brief Generate an new fatbin file with the path in
     /// CudaGpuBinaryFileNames.
-    /// It will add the content of input, to the existing source code, which was
-    /// passed to compileDeviceCode, before.
     ///
-    ///\param [in] input - New source code. The function can select, if code
-    ///       is relevant for the device side. Have to be valid CUDA C++ code.
-    ///\param [in] T - Source of c++ code for variable declaration.
+    /// This interface helps to run everything that the device compiler can run.
+    /// Note that this should be used when there is no idea of what kind of
+    /// input is going to be processed. Otherwise if is known, for example
+    /// only header files are going to be processed it is much faster to run the
+    /// specific interface for doing that - in the particular case - declare().
+    ///
+    ///\param [in] input - The input directly from the UI. Attention, the string
+    /// must not be wrapped or transformed.
+    ///
+    ///\returns true, if all stages of generating fatbin runs right and a new
+    /// fatbin file is written.
+    bool process(const llvm::StringRef input);
```

Comment:
Why is the parameter `const`? Same for the others occurrences in other functions, where parameters are passed by value.

Reply:
`const` is only an optimization. Should I switch to `const std::string &` like the original interfaces of `cling::Intpereter`? `llvm::StringRef` is just a reference. Therefore it is lightweight.

Reply:
replaced `const llvm::StringRef` with `const std::string&`